### ISSUE TITLE
JSON::ParseError - 765: unexpected token

### ISSUE THREAD

**Issue by Adeering** [ankane/searchkick#982]
I get the following error when I use any version of elasticsearch except 2.4 with searchkick:

`JSON::ParserError - 765: unexpected token at '':`

It references my index controller which I currently have set to a simple search everything like this:

`@recipes = Recipe.search "*"`

This is my model:
```
class Recipe < ApplicationRecord
    searchkick
    def search_data
        if attributes.keys.include?("popularity")
            ranking = self["popularity"]
        else
            select_sql = "popularity(view_count, comment_count, submitted_change_count, successful_change_count, created_at, save_count) as popularity"
            popularity = Recipe.select(select_sql).find(id).try(:[], "popularity")
            puts "Popularity: #{popularity}"
        end
        if attributes.keys.include?("most_active")
            most_active = self["most_active"]
        else
            select_sql = "most_active(submitted_change_count, successful_change_count, created_at) as most_active"
            most_active = Recipe.select(select_sql).find(id).try(:[], "most_active")
            puts "Most Active: #{most_active}"
        end
        attrs = attributes.dup
        relational = {
            course_name: course_id(&:name),
            ingredients: recipe_ingredients.map(&:name),
            steps: recipe_steps.map(&:step),
            author_name: user_id(&:displayname),
            cuisines: cuisine_list,
            techniques: technique_list,
            equipments: equipment_list,
            popularity: popularity.to_f,
            most_active: most_active.to_f,
            recipe_changes: recipe_changes.map(&:id)
        }
        attrs.merge! relational
    end
```
I get the same error even when trying to reindex

**Comment by Adeering** [comment]
Anybody got any ideas on this...it is driving me crazy and cant figure it out

**Comment by mikelkew** [comment]
There is a chance that this might be caused by receiving a gzip compressed response from ES - the default HTTP compression behavior changed in 5.0..

If this is the case, you could try using the [faraday_middleware](https://github.com/lostisland/faraday_middleware) gem to ensure that the compressed response is accepted. In our case, we do this in an initializer like so, but customize as necessary for your setup:

```
require "faraday"
require "typhoeus/adapters/faraday"  # remove this if you don't use Typhoeus
require "faraday/middleware"

Searchkick.client = Elasticsearch::Client.new(
  url:                ENV["ELASTICSEARCH_URL"] || "http://localhost:9200",
  transport_options:  {request: {timeout: 30}},
  retry_on_failure:   true
) do |f|
  f.use FaradayMiddleware::Gzip
end
```

**Comment by Adeering** [comment]
Ok, so adding in faraday_middleware gem and setting it to use the Gzip worked on devleopment side.  Ill push an update and make sure to double check the production side as well.  This should get added to the readme setup cause I didn't see this anywhere and have spent the last 3 days narrowing down what causes it and had it narrowed down to when I configure the client with my aws credentials.  Thank you for the response

**Comment by AnkurGel** [comment]
@mikelkew Does this mean that Searchkick client is advertising that it is ready to accept gzipped response? Please correct me if I'm wrong, the server should not be sending gzipped response until the client advertises in way like: `Accept-Encoding: deflate, gzip;q=1.0, *;q=0.5`. If that's indeed the case, `f.use FaradayMiddleware::Gzip` should be default.

**Comment by mikelkew** [comment]
@AnkurGel You're correct, ES shouldn't send a gzipped response without the Accept-Encoding header in the request, however, this problem looks like it _might_ be due to the aws-signers middleware setting a default value for the header. See this [issue](https://github.com/elastic/elasticsearch-ruby/issues/457) and this [pull request](https://github.com/ankane/searchkick/pull/981).

IMHO, Gzip probably should be enabled in most scenarios anyway (or at least in production environments) to reduce the size of the response payloads, especially if frequently expecting large responses.

**Comment by ATMartin** [comment]
We ran into this same problem and resolved it by extracting #981's patch into our initializer:
```
# config/initializers/elasticsearch.rb

Searchkick.client_options = {
  transport_options: {
    headers: { "Accept-Encoding": "" }
  }
}
```

**Comment by ankane** [comment]
For future readers, see #981 for resolution
